posts/2025-05-08-memory_R-jangorecki/memory-R.md

```diff
@@ -0,0 +1,139 @@
+
+# Benchmarking memory usage in R
+
+Profiling memory in R was always not a trivial task.  
+In this post I would like to emphasize that currently popular methods are very much inaccurate, therefore should be used with caution, and more importantly, should not be used to draw conclusions about actual memory usage of an R functions.  
+
+The root cause of inaccuracy of many memory profiling tools in R is that they measure memory allocated by R (including R's C) code. They do not take into account memory allocated using C.  
+
+## Memory allocation in R
+
+Following example should make it very clear.  
+
+Below R chunk is the content of `memtest.R` file.
+```r
+code = "
+  int nx = LENGTH(x);
+  double *y = (double*)(
+    LOGICAL(r_alloc)[0] ?
+      R_alloc(nx, sizeof(double)) : // allocated by R's C
+      malloc(nx * sizeof(double))   // allocated by C
+  );
+  double *xp = REAL(x);
+  // populate y
+  for (int i=0; i<nx; i++)
+    y[i] = xp[i];
+  // do something with y
+  for (int i=1; i<nx; i++)
+    y[i] = y[i-1]+y[i];
+  // sum double array to ensure compiler wont optimize it away
+  double sum = 0.0;
+  for (int i=0; i<nx; i++)
+    sum += y[i];
+  SEXP res = PROTECT(Rf_allocVector(REALSXP, 1));
+  REAL(res)[0] = sum;
+  if (!LOGICAL(r_alloc)[0])
+    free(y);
+  UNPROTECT(1);
+  return res;
+"
+funx = inline::cfunction(signature(x="numeric", r_alloc="logical"), code, language="C")
+set.seed(108)
+x = rnorm(1e8)
+```
+
+## Check equal
+
+First we will ensure that results are the same, no matter if we allocate temporary working memory using R or C:
+
+```sh
+Rscript -e 'source("memtest.R"); funx(x, r_alloc=TRUE)'
+#[1] 1.160649e+12
+Rscript -e 'source("memtest.R"); funx(x, r_alloc=FALSE)'
+#[1] 1.160649e+12
+```
+
+## Memory benchmark using `bench`
+
+Next we will use the most popular, at present, package for profiling memory `bench`:
+
+```sh
+Rscript -e 'source("memtest.R"); bench::mark(funx(x, r_alloc=TRUE))'
+## A tibble: 1 × 13
+#  expression      min median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time
+#  <bch:expr>    <bch> <bch:>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm>
+#1 funx(x, r_al… 577ms  577ms      1.73     763MB     1.73     1     1      577ms
+Rscript -e 'source("memtest.R"); bench::mark(funx(x, r_alloc=FALSE))'
+## A tibble: 1 × 13
+#  expression      min median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time
+#  <bch:expr>    <bch> <bch:>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm>
+#1 funx(x, r_al… 589ms  589ms      1.70        0B        0     1     0      589ms
+```
+
+As we can see in output of `mark` function, `mem_alloc` is reported to be 0B when we use `malloc`, while for `R_alloc` it reports 763MB. The difference we observe here should serve as a warning. If one wants to use `mark` function to draw conculsion about memory usage, one should as well examine source code of the function that is being benchmarked.
```

Comment:
> As we can see in output of `mark` function

As we can see in the output of `mark` function

> If one wants to use `mark` function to draw conculsion about memory usage, one should as well examine source code of the function that is being benchmarked.

If one intends to use the `mark` function to draw conclusions about memory usage, it's crucial to also examine the source code of the function being benchmarked. 

Maybe worth to add that `bench::mark` primarily tracks memory allocations managed by R's memory allocator and doesn't inherently account for memory allocated directly through C functions like `malloc` or `calloc`.